### PR TITLE
Prioritize forks based on upper bounds

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

Given a fork like:

```
pylint < 3 ; sys_platform == 'darwin'
pylint > 2 ; sys_platform != 'darwin'
```

Solving the top branch will typically yield a solution that also satisfies the bottom branch, due to maximum version selection (while the inverse isn't true).

To quote an example from the docs:

```rust
// If there's no difference, prioritize forks with upper bounds. We'd prefer to solve
// `numpy <= 2` before solving `numpy >= 1`, since the resolution produced by the former
// might work for the latter, but the inverse is unlikely to be true due to maximum
// version selection. (Selecting `numpy==2.0.0` would satisfy both forks, but selecting
// the latest `numpy` would not.)
```

Closes https://github.com/astral-sh/uv/issues/4926 for now.

### CODE REVIEW

File: crates/uv-resolver/src/resolver/mod.rs [hunk 1]
```diff
@@ -2843,7 +2843,34 @@ impl Ord for Fork {
...
+                .dependencies
+                .iter()
+                .filter(|dep| {
+                    dep.version
+                        .bounding_range()
+                        .is_some_and(|(_, upper)| !matches!(upper, Bound::Unbounded))
+                })
+                .count();
+
+            self_upper_bounds.cmp(&other_upper_bounds)
```
Comment (charliermarsh) [inline crates/uv-resolver/src/resolver/mod.rs:2872]: Just a heuristic of course, and we don't really have a ton of examples to inform this work.

**Comment by konstin** [comment]
While this is an improvement, it does not fix the transformers example. Fork points for transformers:

* Splitting on pydantic==2.8.2 over typing-extensions
* Splitting on opencv-python==4.10.0.84 over numpy
* Splitting on pandas==2.2.2 over numpy
* Splitting on botocore==1.34.151 over urllib3

I think we need to prioritize `python_version` for that.

**Review by konstin — APPROVED** [review]

**Review by BurntSushi — APPROVED** [review]

**Comment by charliermarsh** [comment]
@konstin - Can you link to that example, and I'll explore it?

**Comment by konstin** [comment]
I'Ve added it in https://github.com/astral-sh/uv/pull/5657

**Comment by charliermarsh** [comment]
Thanks! I'll merge this and then explore that separately.

**Pull request merged**